### PR TITLE
Add support for tagged metrics

### PR DESCRIPTION
**Pull request by chelseakomlo**
After the HTTP endpoint for tagged metrics is merged, extensive QA will be useful.

### CODE REVIEW

File: client/config/config.go [hunk 1]
```diff
@@ -179,6 +179,14 @@ type Config struct {
 	// NoHostUUID disables using the host's UUID and will force generation of a
 	// random UUID.
 	NoHostUUID bool
+
+	// DisableTaggedMetrics determines whether metrics will be displayed via a
+	// key/value/tag format, or simply a key/value format
```
Comment (dadgar) [inline client/config/config.go]: Will the parsing of the config in the agent be a separate PR?

File: client/client.go [hunk 2]
```diff
@@ -1861,58 +1861,132 @@ func (c *Client) emitStats() {
...
 
-// emitHostStats pushes host resource usage stats to remote metrics collection sinks
-func (c *Client) emitHostStats(hStats *stats.HostStats) {
+func (c *Client) setGaugeForMemoryStats() {
 	nodeID := c.Node().ID
-	metrics.SetGauge([]string{"client", "host", "memory", nodeID, "total"}, float32(hStats.Memory.Total))
-	metrics.SetGauge([]string{"client", "host", "memory", nodeID, "available"}, float32(hStats.Memory.Available))
-	metrics.SetGauge([]string{"client", "host", "memory", nodeID, "used"}, float32(hStats.Memory.Used))
-	metrics.SetGauge([]string{"client", "host", "memory", nodeID, "free"}, float32(hStats.Memory.Free))
+	hStats := c.hostStatsCollector.Stats()
```
Comment (dadgar) [inline client/client.go]: Call these in the emitHostStats and pass it in to all the methods?

File: client/task_runner.go [hunk 3]
```diff
@@ -1786,19 +1786,55 @@ func (r *TaskRunner) setCreatedResources(cr *driver.CreatedResources) {
 	r.createdResourcesLock.Unlock()
 }
 
-// emitStats emits resource usage stats of tasks to remote metrics collector
-// sinks
-func (r *TaskRunner) emitStats(ru *cstructs.TaskResourceUsage) {
-	if ru.ResourceUsage.MemoryStats != nil && r.config.PublishAllocationMetrics {
+func (r *TaskRunner) setGaugeForMemory(ru *cstructs.TaskResourceUsage) {
+	if !r.config.DisableTaggedMetrics {
+		labels := []metrics.Label{{"job", r.alloc.Job.Name}, {"task_group", r.alloc.TaskGroup}, {"alloc_id", r.alloc.ID}, {"task", r.task.Name}}
```
Comment (dadgar) [inline client/task_runner.go]: Can we add a field to the task runner struct that holds these baseLabels and create it once when creating the task runner. Avoid making a bunch of new arrays every time

File: client/client.go [hunk 4]
```diff
@@ -1861,58 +1861,132 @@ func (c *Client) emitStats() {
...
 
-	metrics.SetGauge([]string{"uptime"}, float32(hStats.Uptime))
+	if !c.config.DisableTaggedMetrics {
+		labels := []metrics.Label{{"node_id", nodeID}, {"datacenter", c.Node().Datacenter}}
```
Comment (dadgar) [inline client/client.go]: Why don't the others have datacenter?

Comment (dadgar) [reply client/client.go]: Create a base labels on the client as well and where addition labels are needed they can just append and create a new slice?

File: client/client.go (hunk 2 above)
Comment (chelseakomlo) [reply client/client.go]: I was considering creating a separate ClientMetrics object and setting hStats, nodeID, etc as fields. What do you think about that, or should this logic be kept in Client?

Comment (dadgar) [reply client/client.go]: Maybe if we find that we add more than just the stats and nodeID but for just those two I don't think it is worth it.

Comment (chelseakomlo) [reply client/client.go]: Base labels could also be a field- but agree, we can not do it for now and keep it as an option for later.

**Comment by jippi** [comment]
having region and dc in all metrics would be the bomb!

File: client/client.go [hunk 5]
```diff
@@ -2011,18 +2001,26 @@ func (c *Client) setGaugeForAllocationStats() {
 }
 
 // No lables are required so we emit with only a key/value syntax
-func (c *Client) setGaugeForUptime() {
-	hStats := c.hostStatsCollector.Stats()
-	metrics.SetGauge([]string{"uptime"}, float32(hStats.Uptime))
+func (c *Client) setGaugeForUptime(hStats *stats.HostStats) {
+	if !c.config.DisableTaggedMetrics {
+		labels := []metrics.Label{{"datacenter", c.Node().Datacenter}}
```
Comment (dadgar) [inline client/client.go]: Why not baseLabels?

File: client/client.go [hunk 6]
```diff
@@ -187,6 +191,8 @@ func NewClient(cfg *config.Config, consulCatalog consul.CatalogAPI, consulServic
 		serversDiscoveredCh: make(chan struct{}),
 	}
 
+	c.baseLabels = []metrics.Label{{"node_id", c.Node().ID}, {"datacenter", c.Node().Datacenter}}
```
Comment (jippi) [inline client/client.go]: would it be possible to include region too? would make it easier to work with the metrics inside datadog and influx when you have one common metrics storage cross env

Comment (chelseakomlo) [reply client/client.go]: Including the region will take a bit more work and is outside the scope for this PR, but we will definitely keep this in mind for the future. Thanks for the suggestion!

File: client/client.go [hunk 7]
```diff
@@ -284,6 +288,10 @@ func NewClient(cfg *config.Config, consulCatalog consul.CatalogAPI, consulServic
 	// Start collecting stats
 	go c.emitStats()
 
+	// Assign labels at the latest possible moment so the information expected
+	// is ready
+	c.baseLabels = []metrics.Label{{Name: "node_id", Value: c.Node().ID}, {Name: "datacenter", Value: c.Node().Datacenter}}
```
Comment (dadgar) [inline client/client.go]: Probably want this before you start the emitStats go routine right above it

Comment (dadgar) [reply client/client.go]: If anything you could put this in emitStats itself before it starts the loop.

File: client/client.go [hunk 8]
```diff
@@ -1861,58 +1869,116 @@ func (c *Client) emitStats() {
...
 }
 
-// emitHostStats pushes host resource usage stats to remote metrics collection sinks
-func (c *Client) emitHostStats(hStats *stats.HostStats) {
-	nodeID := c.Node().ID
-	metrics.SetGauge([]string{"client", "host", "memory", nodeID, "total"}, float32(hStats.Memory.Total))
-	metrics.SetGauge([]string{"client", "host", "memory", nodeID, "available"}, float32(hStats.Memory.Available))
-	metrics.SetGauge([]string{"client", "host", "memory", nodeID, "used"}, float32(hStats.Memory.Used))
-	metrics.SetGauge([]string{"client", "host", "memory", nodeID, "free"}, float32(hStats.Memory.Free))
+func (c *Client) setGaugeForMemoryStats(nodeID string, hStats *stats.HostStats) {
```
Comment (dadgar) [inline client/client.go]: Please add comments on all the new functions

File: client/client.go [hunk 9]
```diff
@@ -1924,10 +1990,40 @@ func (c *Client) emitHostStats(hStats *stats.HostStats) {
...
+// emitHostStats pushes host resource usage stats to remote metrics collection sinks
+func (c *Client) emitHostStats() {
+	nodeID := c.Node().ID
+	hStats := c.hostStatsCollector.Stats()
+
+	c.setGaugeForMemoryStats(nodeID, hStats)
+	c.setGaugeForUptime(hStats)
+	c.setGaugeForCPUStats(nodeID, hStats)
+	c.setGaugeForDiskStats(nodeID, hStats)
+	c.setGaugeForAllocationStats(nodeID)
```
Comment (dadgar) [inline client/client.go]: This probably shouldn't be guarded by `c.config.PublishNodeMetrics`. This be in client metrics

File: client/client_test.go [hunk 10]
```diff
@@ -121,6 +122,25 @@ func TestClient_StartStop(t *testing.T) {
...
+	t.Parallel()
+	assert := assert.New(t)
+
+	client := testClient(t, nil)
+	if err := client.Shutdown(); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	nodeID := client.Node().ID
+	for _, e := range client.baseLabels {
```
Comment (dadgar) [inline client/client_test.go]: Assert length of base labels != 0

File: client/task_runner.go [hunk 11]
```diff
@@ -1807,3 +1845,15 @@ func (r *TaskRunner) emitStats(ru *cstructs.TaskResourceUsage) {
 		metrics.SetGauge([]string{"client", "allocs", r.alloc.Job.Name, r.alloc.TaskGroup, r.alloc.ID, r.task.Name, "cpu", "total_ticks"}, float32(ru.ResourceUsage.CpuStats.TotalTicks))
 	}
 }
+
+// emitStats emits resource usage stats of tasks to remote metrics collector
+// sinks
+func (r *TaskRunner) emitStats(ru *cstructs.TaskResourceUsage) {
+	if ru.ResourceUsage.MemoryStats != nil && r.config.PublishAllocationMetrics {
```
Comment (dadgar) [inline client/task_runner.go]:
```
func (r *TaskRunner) emitStats(ru *cstructs.TaskResourceUsage) {
        if !r.config.PublishAllocationMetrics {
            return
        }

	if ru.ResourceUsage.MemoryStats != nil {
           ...
        }
        ....
```

File: command/agent/config.go [hunk 12]
```diff
@@ -546,17 +554,19 @@ func DefaultConfig() *Config {
...
+			ClientMinPort:              14000,
+			ClientMaxPort:              14512,
+			Reserved:                   &Resources{},
+			GCInterval:                 1 * time.Minute,
+			GCParallelDestroys:         2,
+			GCDiskUsageThreshold:       80,
+			GCInodeUsageThreshold:      70,
+			GCMaxAllocs:                50,
+			NoHostUUID:                 helper.BoolToPtr(true),
+			DisableTaggedMetrics:       false,
```
Comment (dadgar) [inline command/agent/config.go]: Unneeded since the default value for booleans is false.

File: command/agent/config.go [hunk 13]
```diff
@@ -1032,6 +1042,14 @@ func (a *ClientConfig) Merge(b *ClientConfig) *ClientConfig {
 		result.NoHostUUID = b.NoHostUUID
 	}
 
+	if b.DisableTaggedMetrics {
```
Comment (dadgar) [inline command/agent/config.go]: Merge test: https://github.com/hashicorp/nomad/blob/master/command/agent/config_test.go#L71
